ledger/participant-integration-api/src/main/scala/platform/store/backend/common/EventStorageBackendTemplate.scala

```diff
@@ -0,0 +1,464 @@
+          |  event_kind != 0 AND -- we do not want to fetch divulgence events""".stripMargin,
+        List(
+          "transactionId" -> transactionId
+        ),
+      ),
+      rowParser = rawFlatEventParser,
+      selectColumns = selectColumnsForFlatTransactions,
+      witnessesColumn = "flat_event_witnesses",
+    )(
+      limit = None,
+      fetchSizeHint = None,
+      filterParams = filterParams,
+    )(connection)
+  }
+
+  override def transactionTreeEvents(
+      rangeParams: RangeParams,
+      filterParams: FilterParams,
+  )(connection: Connection): Vector[EventsTable.Entry[Raw.TreeEvent]] = {
+    events(
+      columnPrefix = "",
+      joinClause = "",
+      additionalAndClause = (
+        """  event_sequential_id > {startExclusive} AND
+          |  event_sequential_id <= {endInclusiveSeq} AND
+          |  event_kind != 0 AND -- we do not want to fetch divulgence events""".stripMargin,
+        List(
+          "startExclusive" -> rangeParams.startExclusive,
+          "endInclusiveSeq" -> rangeParams.endInclusive,
+        ),
+      ),
+      rowParser = rawTreeEventParser,
+      selectColumns =
+        s"$selectColumnsForTransactionTree, ${eventStrategy.columnEqualityBoolean("event_kind", "20")} as exercise_consuming",
+      witnessesColumn = "tree_event_witnesses",
+    )(
+      limit = rangeParams.limit,
+      fetchSizeHint = rangeParams.fetchSizeHint,
+      filterParams,
+    )(connection)
+  }
+
+  // TODO the ordering of the results here is based on node_index in the previous implementation. current implementation uses event_sequential_id. Are these the same? (if not perhaps flatTransaction needs adapting too?)
+  override def transactionTree(
+      transactionId: TransactionId,
+      filterParams: FilterParams,
+  )(connection: Connection): Vector[EventsTable.Entry[Raw.TreeEvent]] = {
+    import com.daml.platform.store.Conversions.ledgerStringToStatement
+    events(
+      columnPrefix = "",
+      joinClause = """JOIN parameters ON
+          |  (participant_pruned_up_to_inclusive is null or event_offset > participant_pruned_up_to_inclusive)
+          |  AND event_offset <= ledger_end""".stripMargin,
+      additionalAndClause = (
+        """  transaction_id = {transactionId} AND
+          |  event_kind != 0 AND -- we do not want to fetch divulgence events""".stripMargin,
+        List(
+          "transactionId" -> transactionId
+        ),
+      ),
+      rowParser = rawTreeEventParser,
+      selectColumns =
+        s"$selectColumnsForTransactionTree, ${eventStrategy.columnEqualityBoolean("event_kind", "20")} as exercise_consuming",
+      witnessesColumn = "tree_event_witnesses",
+    )(
+      limit = None,
+      fetchSizeHint = None,
+      filterParams,
+    )(connection)
+  }
+
+}
+
+trait EventStrategy {
+  def filteredEventWitnessesClause(
+      witnessesColumnName: String,
+      parties: Set[Ref.Party],
+  ): (String, List[NamedParameter])
```

Comment:
I assume each method is responsible for finding globally unique names for the named parameters? Would anorm complain if you accidentally defined the value of a named parameter twice?

Reply:
Oh yes anorm will complain then! I added comments to clarify this behavior

Reply:
Correction: anorm won't complain. I made a comment about that, will improve this in the next PR